test/test_nn.py

```diff
@@ -532,7 +532,8 @@ def _backward_criterion(self, criterion, input, target, gradOutput=None, extra_a
     def _zero_grad_parameters(self, module):
         for p in module.parameters():
             if p.grad is not None:
-                p.grad.data.zero_()
+                with torch.no_grad():
+                    p.grad.zero_()
```

Comment:
A little confused here. You said:

> If in_place_operation_() only updates the data inside the tensor (such as zeros_()), such operation will still work properly; if the in-place operation changes the size, stride or the storage pointer inside the TensorImpl (e.g. resize_ / resize_as_ / set_ / transpose_), such operation on tensor.data or tensor_detached will no longer update the tensor.

So why do you have to say `torch.no_grad()` here?

Reply:
For sparse tensors, `indices` and `values` are also considered as a tensor's metadata, and changing those two fields on a derived tensor (i.e. tensor from `.data`) will not update the original tensor.

Here, when `p.grad` is a sparse tensor, `p.grad.data.zero_()` changes `p.grad.data`'s underlying `indices` and `values` tensor, but `p.grad`'s `indices` and `values` tensor are not updated, which causes test error.

Reply:
This seems a bit fishy.  For example, couldn't you implement zero_() by not changing the shape of the metadata and have this working?  Obviously this would not be very efficient.

But I don't really understand what our guarantees are for SparseTensors with autograd -- @SsnL can you chime in?

Reply:
Behavior of some sparse tensor operations with/without `allow_tensor_metadata_change_` flag:

`sparse_tensor.zero_()`
```python
>>> import torch
>>> i = torch.tensor([[0, 1, 1], [2, 0, 2]])
>>> v = torch.tensor([3., 4., 5.])
>>> st = torch.sparse.FloatTensor(i, v)
>>> dst = st.data

# Without the allow_tensor_metadata_change_ flag:
>>> dst.zero_()   # Note that `dst` and `st` have different values now
>>> dst
tensor(indices=tensor([], size=(2, 0)), values=tensor([], size=(0,)), size=(2, 3), nnz=0, layout=torch.sparse_coo)
>>> st
tensor(indices=tensor([[0, 1, 1], [2, 0, 2]]), values=tensor([3., 4., 5.]), size=(2, 3), nnz=3, layout=torch.sparse_coo)

# With the allow_tensor_metadata_change_ flag:
>>> dst.zero_()   # This operation is disallowed, because it breaks consistency between `dst` and `st`
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: resize_and_clear_ is not allowed on Tensor created from .data or .detach()
```

`sparse_tensor.copy_(st)`
```python
>>> import torch
>>> i = torch.tensor([[0, 1, 1], [2, 0, 2]])
>>> v = torch.tensor([3., 4., 5.])
>>> st = torch.sparse.FloatTensor(i, v)
>>> dst = st.data

# Without the allow_tensor_metadata_change_ flag:
>>> i2 = torch.tensor([[0, 1, 1], [2, 0, 2]])
>>> v2 = torch.tensor([10., 10., 10.])
>>> st2 = torch.sparse.FloatTensor(i2, v2)
>>> dst.copy_(st2)  # Note that `dst` and `st` have different values now
>>> dst
tensor(indices=tensor([[0, 1, 1], [2, 0, 2]]), values=tensor([10., 10., 10.]), size=(2, 3), nnz=3, layout=torch.sparse_coo)
>>> st
tensor(indices=tensor([[0, 1, 1], [2, 0, 2]]), values=tensor([3., 4., 5.]), size=(2, 3), nnz=3, layout=torch.sparse_coo)

# With the allow_tensor_metadata_change_ flag:
>>> dst.copy_(st2)   # This operation is disallowed, because it breaks consistency between `dst` and `st`
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: resize_ is not allowed on Tensor created from .data or .detach()
```

`sparse_tensor.add_(st)`
```python
>>> import torch
>>> i = torch.tensor([[0, 1, 1], [2, 0, 2]])
>>> v = torch.tensor([3., 4., 5.])
>>> st = torch.sparse.FloatTensor(i, v)
>>> dst = st.data
>>> st2 = torch.sparse.FloatTensor(torch.tensor([[0, 1, 1], [2, 0, 2]]), torch.tensor([10., 10., 10.]))

# Without the allow_tensor_metadata_change_ flag:
>>> dst.add_(st2)   # Note that `dst` and `st` have different values now
>>> dst
tensor(indices=tensor([[0, 1, 1], [2, 0, 2]]), values=tensor([13., 14., 15.]), size=(2, 3), nnz=3, layout=torch.sparse_coo)
>>> st
tensor(indices=tensor([[0, 1, 1], [2, 0, 2]]), values=tensor([3., 4., 5.]), size=(2, 3), nnz=3, layout=torch.sparse_coo)

# With the allow_tensor_metadata_change_ flag:
>>> dst.add_(st2)   # This operation is disallowed, because it breaks consistency between `dst` and `st`
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: set_indices_and_values_unsafe is not allowed on Tensor created from .data or .detach()
```

So the consistency-breaking changes can be prevented by the `allow_tensor_metadata_change_` flag.

Reply:
A better fix here might be the following, to reflect the different semantics for dense vs. sparse:
```python
if p.grad is not None:
    if p.is_sparse:  # `zero_()` on sparse tensor changes tensor metadata, and it should only be performed on the original tensor
        p.grad.zero_()
    else:
        p.grad.data.zero_()
```